formats.kdl: init

KiaraGrouwstra opened the pull request:
## Description of changes

Adds `pkgs.formats.kdl`, a Nix representation of the [KDL document language](https://kdl.dev/) using the standard `pkgs.formats` format.

Also inits `json2kdl`, a custom-built application which is required for the core functionality of `pkgs.formats.kdl`.

Closes #198655.
Superseded #295211 to incorporate feedback (/cc @feathecutie @eclairevoyant @Stunkymonkey @SuperSandro2000 @infinisil @h7x4 @sodiboo @Lyndeno).

## Things done

* Built on platform(s)
  * [x]  x86_64-linux
  * [ ]  aarch64-linux
  * [ ]  x86_64-darwin
  * [ ]  aarch64-darwin
* For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  * [ ]  `sandbox = relaxed`
  * [ ]  `sandbox = true`
* [ ]  Tested, as applicable:
  * [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  * and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  * or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  * made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
* [x]  Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
* [x]  Tested basic functionality of all binary files (usually in `./result/bin/`)
* [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  * [ ]  (Package updates) Added a release notes entry if the change is major or breaking
  * [ ]  (Module updates) Added a release notes entry if the change is significant
  * [ ]  (Module addition) Added a release notes entry if adding a new NixOS module
* [x]  Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

KiaraGrouwstra commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 412029`
Commit: `df626b52cbe48208050d8b480c073c4dc9f58dc3`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>json2kdl</li>
  </ul>
</details>

KiaraGrouwstra commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 412029`
Commit: `ee6046a1f116e8cd8976ba8fb8a1b2960e12935c`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>json2kdl</li>
  </ul>
</details>

SuperSandro2000 commented:
Ui, nice, thanks! But I don't think I can add much valuable review here

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2401

KiaraGrouwstra commented:
@flokli as a niri user, would you mind reviewing as well? :see_no_evil:

bloxx12 commented on pkgs/pkgs-lib/formats.nix:
```diff
@@ -895,4 +895,173 @@ rec {
+    {
+
+      type = (
+        with lib.types;
```
Can you drop this? Most, if not all of the types are already inherited at the top of the file.

sodiboo commented on nixos/doc/manual/development/settings-options.section.md:
```diff
@@ -364,6 +364,67 @@ have a predefined type and string generator already declared under
     and returning a set with [CDN](https://github.com/dzikoysk/cdn)-specific
     attributes `type` and `generate` as specified [below](#pkgs-formats-result).
 
+`pkgs.formats.kdl` { }
+
+:   A function taking an empty attribute set (for future extensibility)
```
Hello. It is the future. There should be an option to choose which version of KDL to output.

There are exactly two versions, KDL v1.0.0 and KDL v2.0.0. It would be nice if we could make this *explicitly specified* and *required* before this format is merged.

sodiboo replied:
you added the version like i asked, but forgot to update this part of the manual

```suggestion
`pkgs.formats.kdl` { `version` }

:   A function taking a version number (which must be exactly `1`, for future extensibility)
```

sodiboo commented on pkgs/pkgs-lib/formats.nix:
```diff
@@ -988,4 +990,174 @@ optionalAttrs allowAliases aliases
+    {
+      version,
+    }:
+    lib.mkAssert (version == "1.0.0") "json2kdl so far supports only KDL 1.0.0" {
```
(nitpick?) should this really be a string? i'd imagine it should be an int. `version = 1` || `version = 2` (the latter not being supported right away, and doesn't have to be)

in particular, KDL formally has specs in Semantic Versioning, 1.0.0 and 2.0.0, to signify major/breaking changes clearly, but like, *they don't really have minor changes*. The v1.0.0 spec URL is just [`/spec-v1`](https://kdl.dev/spec-v1/) and in many other places [like here](https://github.com/kdl-org/kdl-rs/blob/439aa63bfc595432e6dc6b0fdd708c364628ab18/src/lib.rs#L108-L115) it's just informally referred to as "kdl v1" and "kdl v2".

moreover, if there are non-breaking non-major version-bumping changes, there's not really any reason for us to distinguish them. like [this thing](https://github.com/kdl-org/kdl/pull/513) might lead to a KDL v2.1.0 down the line, but it's purely additive syntax sugar. the data model is identical and the syntax is backwards-compatible. the reference implementation would still just refer to them as v1 and v2.

sodiboo reviewed (approved):
i'm quite happy with how this looks, and it suits my needs for e.g. `niri-flake`.

KiaraGrouwstra commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 412029`
Commit: `9cf39599b58358bc395e6f58735a278645d4b5c6`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>json2kdl</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2428

Stunkymonkey reviewed (commented):
could we please have some tests in `pkgs/pkgs-lib/tests/formats.nix`

eclairevoyant commented on pkgs/pkgs-lib/formats.nix:
Evidently this file needs `nixfmt-rfc-style` rerun on it to pass CI (may need a rebase on `master` if that formatter changed in the meantime)

wolfgangwalther reviewed (dismissed):
sry, mis-clicked - consider my previous review blocking!

bloxx12 commented on pkgs/by-name/js/json2kdl/package.nix:
```diff
@@ -0,0 +1,29 @@
+  fetchFromGitHub,
+}:
+
+rustPlatform.buildRustPackage rec {
```
```suggestion
rustPlatform.buildRustPackage (finalAttrs: {
```

bloxx12 commented on pkgs/by-name/js/json2kdl/package.nix:
```diff
@@ -0,0 +1,29 @@
+      kiara
+    ];
+  };
+}
```
```suggestion
})
```

eclairevoyant commented on pkgs/by-name/js/json2kdl/package.nix:
```diff
@@ -0,0 +1,31 @@
+  meta = {
+    description = "Program that converts JSON files to KDL";
+    homepage = "https://github.com/AgathaSorceress/json2kdl";
+    license = lib.licenses.unfree; # NVPLv7+
```
Please add [the nvpl](https://git.pixie.town/thufie/npl-builder/src/branch/main/nvpl.md) to lib.licenses, which makes it easier for other packages to use it and actually tracks it as its own license (for allowing certain unfree licenses, for example)

eclairevoyant commented on pkgs/by-name/js/json2kdl/package.nix:
```diff
@@ -0,0 +1,31 @@
+  src = fetchFromGitHub {
+    owner = "AgathaSorceress";
+    repo = "json2kdl";
+    rev = "v${finalAttrs.version}";
```
nit, but since I'm already requesting changes...
```suggestion
    tag = "v${finalAttrs.version}";
```

KiaraGrouwstra commented:
given the unfree license complicates the test, to unblock the package i've split that one off to #421700.

joshprk commented:
I've written an equivalent MIT-licensed alternative if necessary. Not intended to replace as the maintainers of json2kdl haven't responded yet, but just offering alternatives just in case.

https://github.com/joshprk/jsonkdl/

KiaraGrouwstra commented:
i've now updated the PR to use jsonkdl over json2kdl

KiaraGrouwstra committed:
formats.kdl: init

Update pkgs/pkgs-lib/formats.nix

Co-authored-by: sodiboo <37938646+sodiboo@users.noreply.github.com>

Update nixos/doc/manual/development/settings-options.section.md

Co-authored-by: sodiboo <37938646+sodiboo@users.noreply.github.com>

use submoduleWith

assert version

Aleksanaa commented on nixos/doc/manual/development/settings-options.section.md:
```diff
@@ -364,6 +364,67 @@ have a predefined type and string generator already declared under
+:   A function taking a KDL version (`version == 1 || version == 2`)
+    and returning a set with [KDL](https://kdl.dev/)-specific attributes `type`,
+    `lib` and `generate` as specified [below](#pkgs-formats-result).
+
```
We don't mention `generate` takes a list of "KDL nodes" (and what is it) here. It would be confusing if people try to build a single node with it.

KiaraGrouwstra commented:
@bloxx12 i think i've addressed the reviewers' suggestions here, so as far as i'm concerned this is ready - further reviews welcome

nezia1 commented:
Hey, after trying to use it to generate zellij configuration, I noticed something that does not look right: 

This: ```nix
      ".config/zellij/config.kdl".source = settingsFormat.generate "zellij-config" [
        (node "simplified_ui" null [true] {} [])
      ];
```

Gives me this: ```kdl
simplified_ui #true {
}
```

I would expect the aforementioned nix code to generate this instead:
```kdl
simplified_ui # true;
```
Passing in a null value to `children` does not seem to work either, and makes `jsonkdl` crash with: ```console
╰─󰁔 nix log /nix/store/7l8zdcvnw9ljj41japjfyqdcn9sbk7li-zellij-config.drv
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/7l8zdcvnw9ljj41japjfyqdcn9sbk7li-zellij-config.drv^*'
Error: Conversion(InvalidStructure("document root must be json array"))
```

sodiboo commented:
> I would expect the aforementioned nix code to generate this instead:
> 
> ```kdl
> simplified_ui # true;
> ```

With a space in between `#` and `true`? That's invalid KDL. Is this a typo, and you only meant to show the difference in explicitly-empty vs elided children (which are identical, as you've noticed)?

nezia1 commented:
> > I would expect the aforementioned nix code to generate this instead:
> 
> > 
> 
> > ```kdl
> 
> > simplified_ui # true;
> 
> > ```
> 
> 
> 
> With a space in between `#` and `true`? That's invalid KDL. Is this a typo, and you only meant to show the difference in explicitly-empty vs elided children (which are identical, as you've noticed)?

That was indeed a typo, and yes that is exactly what I meant. Just realized afterwards that they are identical

sodiboo commented:
I do agree that it would be nicer if there weren't any curlies there.

At the level of `kdl-rs` (which `jsonkdl` uses), there is a distinction between elided and empty children (`children: Option<KdlDocument>`), because it deals with concrete syntax trees and needs to know whether there are any curly braces physically. When generating the kdl document in `jsonkdl`, if `children` is an empty array, then the children are set to an empty document, but if the key is missing from the json document, then the children are not set all. @joshprk is this intended behaviour? because to be honest, that's totally sane and reasonable of `jsonkdl` to do. If it always generated elided children, when the input was an empty array, then there would be no way to *explicitly* get one or the other. Maybe that's fine? `jsonkdl` doesn't really expose any other concrete syntax choices; but it also makes sense as-is.

At the nixpkgs level, we should absolutely not expose this distinction: it does not matter a single bit to the meaning of the kdl document, and we likewise don't expose the ability to insert comments. But if we assume jsonkdl wants that behaviour, it might be reasonable to remove the `children` key from whatever we pass to it, if a node has an empty childrens' array.

also none of this matters lol. it makes zero difference. i'm bikeshedding over the *formatting* of a *generated file*

and for that matter, either way is wrong! in e.g. niri, there are many nodes that represent "sections" of the config. such as the `environment {}` section to declare environment variables. by eliding the curlies, you get a config that just mentions the single word `environment` somewhere. the application-specific way to format this is *with* the curlies (or the node can be omitted entirely), which reads as "oh, no environment variables are set". but just `environment` without curlies is kind of baffling.

nezia1 commented:
I apologize for the additional bike shedding, but I have concerns on how this would be incorporated into modules. I assume that, in order to abide by [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md), every module would have to implement their own little abstraction over kdl. Wouldn't that be a bit of an issue, considering that nix attrsets tend to be a little incompatible / ambiguous with kdl syntax?

An example of this would be adding properties to a node; Intuitively I would think of doing something like:

```nix
input = {
  name = {
    prop1 = "foo";
    prop2 = "bar";
  };
};
```

Problem is, how do we know whether this is properties or two embedded nodes here?

I don't think we can expect consumers of the module to bind `kdl = pkgs.formats.kdl {}` in their own files in order to just use `lib.node`, as that would be unwiedly and go against the abstraction provided by the module system. I'm unsure how this could be implemented, or if there should be a written standard on it in the documentation, but I do think that this is something that needs to be addressed before the PR gets merged.

KiaraGrouwstra commented:
@nezia1:
> I have concerns on how this would be incorporated into modules. I assume that, in order to abide by [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md), every module would have to implement their own little abstraction over kdl.

Under RFC42, I wonder if we could add such conversions in a way that would not tie them to applications' settings.
Given my limited bandwidth tho, I'd rather defer to others on such improvements - I wonder if perhaps this could perhaps be handled in a follow-up PR.

sodiboo commented:
I think that as a first step, we don't need a super sophisticated merging solution *yet*. It sucks that KDL settings modules will not be "mergeable" by default. We could perhaps recommend creating a a set of minimal scaffolding options, with inspiration from modules that otherwise have their own custom freeform text shaped configs. `services.caddy` has some helpers for the toplevel structure of their configuration: it lets you define the global config, some ACME service integration, and separating the virtual hosts; but if you even want to do *any* routing you *have* to use the unshaped `extraConfig`. There's even the option to opt out entrely and just provide your own `configFile` directly.

The KDL format is really more like `extraConfig` with `types.lines` than any "settings" shaped format. We should likewise always use the `config` terminology to refer to KDL-shaped configuration options.

And then, in a *separate* pull request, we should create an abstraction for the KDL type: call it `extensible-kdl` or whatever. A useful way i would think this can work is if we have every definition define some *transformation*, so instead of just putting a node, we'd have a more specialized thing like `find-equal-and-merge-children-or-new-node` (could perhaps be called `section`), and the first definition like this would create a new node, but the second one would merge into the existing node.

Maybe the "document scope" could have a set of default merging rules for certain types of nodes, (and every list of children would also introduce a new "document scope")

However that will look, i think it's future work. It's not directly related to this PR. I'd love if we can first focus on just standardizing the *representation for a KDL document in NIx and how to serialize it*, without getting distracted with also making it luxurious to *use in many modules*.

This PR doesn't define a proper "settings format": it's just the serialization format. `formats.kdl` is one step higher and nicer to use than a raw string, but it is still the "raw" config format: we shouldn't pretend like it's supposed to be the final step in RFC 0042 compliance for KDL-configured applications.

sodiboo commented:
> So you suggest that every module should implement their own mapping?

For now, with just the kdl format, yes, every module making *direct use of it* would provide their own mapping. I'm not super well-informed of all the potential packages that want to make use of this, but for e.g. niri i already maintain a comprehensive `settings`-like mapping. Its ideal output is a representation of a KDL document, exactly the *input* to this PR. Internally, i manipulate the *structured representation of the KDL document*, so it's also necessary to have such a higher level representation of a KDL document. I also maintain a serializer in that repo, which can be eliminated entirely if there is a proper format definition in nixpkgs.

> But this would mean that for now, it'd be easier to just provide a `types.lines` option and write it verbatim?

The KDL format behaves *similarly* to `types.lines`, yes, but i don't think it's "better" to use that one: if you do this, you will be exposed to nuances of the KDL syntax. `formats.kdl` abstracts the syntax and formatting (including differences between v1/v2). It also makes correctly inserting *strings* way more painful: if you wanted to insert stuff inside of e.g. [niri's `custom-shader` option](https://github.com/YaLTeR/niri/wiki/Configuration:-Animations#custom-shader) as a string literal, you'd have to think about *three* nested string syntaxes (GLSL-in-KDL-in-Nix), as opposed to just two if we support serializing KDL documents in nixpkgs.

> Fair. I apologize for bike shedding again, I thought that this would be necessary in order to get things moving, but it'd make more sense to have it in a separate PR.

I do think it's useful discussion; i don't want to discourage it altogether, it's just *definitely not* what this PR needs.

> The thing that bothers me however is that this seems to not be an issue for the other config formats; their `generate` functions are able to be used for RFC 0042 compliance just fine. Is this just a byproduct of being easier formats to map one-to-one or is that actually the point of `pkgs.formats`? This is unclear to me and I don't think I was able to find information on this.

This is a byproduct of those formats being easier to map to Nix. I've heard a hypothesis that "every serialization format is either JSON or XML".

Nix is very easy to map to JSON-like formats, which is almost every single one of them, because that's the native representation of data structures in every programming language (except like, uiua). YAML is just a remix of JSON, so is TOML. INI is just JSON with slight restrictions; systemd units are like INI (which is like JSON). And Nix's data model is also basically the same as JSON (except we can also represent infinitely nested objects; serialization of these will *fail*). The thing that these formats have in common is that their primary mode of composition is a *set of attributes* with unique named keys. That's why they're so easy to merge, as well, because this is the same composition primitive that `evalModules` requires.

XML is fundamentally based on the "node", which may have certain structured attributes, but then have a *list* of children. There's no uniqueness requirement. Concatenation *is* the way to merge them. If you have multiple parts of a config, you concatenate them to get the full config. The problem is that `evalModules` allows different places of evaluation to define the *same* part of the config. Merging is only an issue when you have two entirely different spots in the niri config defining e.g. `input {}` configuration. With JSON-like composition, the module system *has* to pick which value to choose, or if it's an attrset defined in two places with the same priority, there is *only one way to merge them*: key-by-key. with XML-like composition, the module system *doesn't know* that there is a conflicting value: There is nothing inherent to the config format that defines this. It's also *application-specific* whether the correct course of action is to "merge" with a previous node (`input {}` section), or always concatenate (`spawn-at-startup` entry).

KDL is like XML. Also, nginx configuration, and Caddyfiles, are both like XML They also have a much more nuanced syntax than KDL, which is why KDL is a very great fit for being serialized in nixpkgs and they're not.

I *assume* that the point of `pkgs.formats` *was* originally to define serialization primitives for Nix representations of data formats, which could be used to represent a mergeable configuration in its entirety, without needing to map stuff to Nix. It's also in `pkgs` because the generation functions can depend on the rest of nixpkgs (e.g. our generation depends on the `jsonkdl` package (to encode KDL-in-JSON and then convert that to actual-KDL). However, i don't *think* that's what it *ought to represent*: it should just be serialization of those formats. The fact that JSON-like formats also represent a fully mergeable *configuration* as-is, makes everyone's lives a lot easier. But that is *really just a coincidence* of those formats being *genuinely easier to deal with*. It shouldn't be the requirement that every format in `pkgs.formats` is equally easy to merge with overrides.

Also, our issues are *not* unique to "XML-like" formats, even. For instance, in systemd configuration, there currently is *no way* in NixOS to remove a single `After=` bound from a service. The only way to "remove" a bound that is already set, is to *overwrite all the bounds*. This is arguably "wrong", because you should be able to create negative overrides: that's the *entire point of the module system*.

The problem is with merging lists: If lists had more nuanced merging option than `lib.mkOverride` and `lib.mkOrder`, then KDL would automatically benefit from all those same options.

A better name than "JSON-like" and "XML-like" formats is "set-oriented" and "list-oriented". Neither are inherently impossible to deal with, and *both* actually support both kinds of syntax (i said systemd was JSON-like, but `After=` is a list!) and it's *just the lists* that are painful to deal with. The reason KDL has huge usability issues this way, is because *the entire config is a list at the top level*, so it's *entirely concatenative*.

---

With all this out of the way, maybe a method we could use to *force* module authors to write more "sound" KDL modules, is to actually make the main KDL type `mergeOneOption`? this prevents users from shooting themselves in the foot by accidentally declaring the same section twice in different files, and then generating a completely wrong-shaped configuration. Some kind of `extensible-kdl` settings format would eventually lift this restriction.

Making it `mergeOneOption`-only, would also at least fit my `niri-flake` use case, because as explained, my module ideally outputs natively a fully-formed KDL document that shouldn't be modified any further.

At the level of niri, i think a sufficient configuration to expose is just a single `config` which is a `mergeOneOption` KDL document. There aren't any other nixpkgs or home-manager modules which would want to modify it, except, like, `stylix`. :( (but this theming is fairly trivial and such projects could have a template for inserting their values into your config). That is totally usable in the real world, by actual users. [The author of this PR does exactly that, in present day](https://codeberg.org/kiara/dots/src/commit/5f546c2838e4617de117df8a8d0039b369196053/system/niri/config.nix). [Eons ago, this is what my configuration used to do](https://github.com/sodiboo/system/commit/bb3962e7cfdb371b4039ae30c2cde30ea09a3d8c). [Here's another config like that](https://github.com/dscv101/andromeda/blob/main/andromeda-niri/config.nix). [This one, too](https://github.com/veryfastman/nix-config/blob/480ddf2f0ebd0c047371049a33e19a061c2df3a2/machines/laptop/niri_conf.nix).

Proper merging of multiple definitions of the same part of the same config, *can* wait, because it's only actually a huge concern on "core" dependencies, which are used in *several other places throughout nixpkgs*. I'm not aware of any other KDL-configured application like that.

KiaraGrouwstra commented:
> I think that we agree that KDL would belong in the non-nix representable category

now, the format `jsonkdl` expects that the `node` function here converts to is just json/nix, meaning we can represent it as such - tho it may be more verbose than using the helper functions here.

to qualify your statement then, would it be closer maybe to say the challenges lie in merging specifically?

nezia1 commented:
> > I think that we agree that KDL would belong in the non-nix representable category
> 
> now, the format `jsonkdl` expects that the `node` function here converts to is just json/nix, meaning we can represent it as such - tho it may be more verbose than using the helper functions here.
> 
> to qualify your statement then, would it be closer maybe to say the challenges lie in merging specifically?

It would yes. Especially since, how we mentioned before, we also need to keep the order intact (and I don't think this is something that can be done with nix attrsets), which means that we'd pretty much have to use a list as the input, as we're doing right now. If we cannot take an attrset as input, can we really make it a `pkgs.formats` entry?

Especially since we might have to add extra underscore attributes to disambiguate, I feel like it just goes against what the manual states.

sodiboo commented:
> This actually seems to not be the case (and this might be where our disagreements lie); Here's what I found on this, since the NixOS manual mentions `pkgs.formats` under [Options for Program Settings](https://nixos.org/manual/nixos/stable/#sec-settings-options):
> 
> > Nix-representable ones: These can trivially be mapped to a subset of Nix syntax. E.g. JSON is an example, since its values like {"foo":{"bar":10}} can be mapped directly to Nix: { foo = { bar = 10; }; }. Other examples are INI, YAML and TOML. The following section explains the convention for these settings.
> > Non-nix-representable ones: These can’t be trivially mapped to a subset of Nix syntax. Most generic programming languages are in this group, e.g. bash, since the statement if true; then echo hi; fi doesn’t have a trivial representation in Nix.
> 
> I think that we agree that KDL would belong in the non-nix representable category, which would therefore contradict the point of `pkgs.formats` (as the languages need to be directly serializable from a nix attrset).

KDL is Nix-representable. I did bring up nginx and Caddyfiles as examples of "perhaps they are similar to KDL" ("XML-based"), but these are more like programming languages, and have many different kinds of statements/directives. I was mistaken to bring them up as equivalents, because KDL is *way simpler* and is *just a document format*. Its data model is dead simple, and does have a trivial representation in Nix. The fact that the Nix representation does not play super nicely with the module system *is* unfortunate. But it is still definitely Nix-representable.

> The subsequent section on Nix-representable formats, mentions pkgs.formats:
> > By convention, formats like this are handled with a generic settings option, representing the full program configuration as a Nix value. The type of this option should represent the format. The most common formats have a predefined type and string generator already declared under pkgs.formats
> 
> Considering how complex kdl may be, I still think this PR is necessary, but I also think that it might not belong under `pkgs.formats`. However I am unsure about namespaces, as this still needs to be under `pkgs` (because of `jsonkdl`). Would `writers` perhaps be a better fit for this?

As said, i already believed the original intention of `pkgs.formats` was "fully mergeable settings formats". I don't think that's how it should be; i think `kdl` fits in there fine, and as said, it's *already sufficient* for most applications (this is completely okay for real-world users of niri, for instance). We can already "represent the full program configuration as a Nix value".

I also recognize that there *is* a significant difference in usability between what's already in `pkgs.formats` and our KDL format. I wouldn't be opposed to putting it in a different namespace. I do think it is unnecessary, but it makes basically zero difference to me.

> Considering how complex kdl may be,

KDL is not complex. It's got a dead simple data model that hasn't changed since v1.

sodiboo commented:
By the way, what do we think of making the type exposed by this PR `mergeUniqueOption`? I think that would make it *clear* that it is not really an end-state answer to RFC 0042, but just a building block.

Most of my - admittely somewhat verbose - discussion above can be summarized as "Yes, but please, *this thing* doesn't need to fundamentally change, it's already usable and *needs to exist*", however i ended up bringing up one important usability point in all that thinking:

I believe the default merging method of concatenating separate definitons is just straight up a footgun and we should avoid it. In the current PR, this would look like changing both instances of `listOf node` into `types.uniq (listOf node)`.

those two instances (in `children =` as well as `valueType =`) can/should probably just be merged into a single `document` type, and that should have its options hidden: there's no need to expose the toplevel `type`,`arguments`,`properties` submodule options; these are really already defined as part of our KDL format. and, they should "technically" just be implementation details of the representation between `kdl.lib.node` and `kdl.lib.generate`. of course, we're going to keep their shape stable, but we don't need to expose that shape in *every appication's options in the manual*. it should just link to the one source for documentation on.

### making the kdl type more precise

i also think it's probably a mistake exposing the full module system in the values of those types? ideally, we want an array of `kdl.lib.node` to be a *fully formed* KDL document, that `kdl.type` will *validate*, right? does this make sense? i don't think that "typecheck" should be allowing e.g `lib.mkMerge` on its children, because really that just means "concatenate" (whereas merging is application-specific); so the user should be using primitives like `++` and `concatLists` here. therefore, i think it makes sense to e.g. apply the `uniq` bound on basically everything in a node. this restricts what the end user can do with this.

i'd love for the kdl type to basically only accept the same things as `jsonkdl`. *except* that it also validates the input at nix-eval-time and can provide better type errors in Nix evaluation. it should also normalize a "partial" KDL document (something like `[{name = "node";}]`) into a fully formed one with every option defaulted (like `[{name = "node"; type = null; arguments = []; properties = {}; children = [];}]`). this normalization is already done, but it should be *the only observable effect* of merging the KDL document.

using `uniq` on everything gets most of the way there, but still allows the elements of the list to be `lib.mkIf` to "eliminate" a node. again, i think that this probably a mistake. i think that the KDL document *typecheck* should not allow conditionals like this (just use regular language primitives; this is all being eval'd and `toJSON`ed in one go anyway). we could make a simpler list type, that suits the KDL needs, and only accepts a flat list of KDL nodes.

in fact, i did exactly that, [here's what i mean by all this](https://github.com/sodiboo/nixpkgs/commit/f4d835044436bfb090df0574717db520b4e8118f). (note: you can still mess around with `lib.mkIf` in the types, names, and values-with-explicit-type-annotation; but this can only ever be passed to `mergeEqualOption` so you can't get any "weird" results; note 2: i made sure you can't do arbitrary module evaluation, by making every submodule check `isAttrs` and `shorthandOnlyDefinesConfig=true`, which is implicit for `types.submodule`)

---

ultimately, the most important thing to me, is just that the *very toplevel* type must be `mergeUniqueOption`. the rest of it *ought to* be "locked down" too so it's more difficult to use incorrectly; we shouldn't allow weird module stuff in it.

### making the kdl type more ergonomic

the above solution has one major drawback: the absolute toplevel list is now strict in its "final" length; which *is* problematic. because, if you want to, for instance, add an optional node, then the condition for adding that node must now be evaluated to get the fact that the document as a whole is a list. so it *is* actually not as lazy as one would prefer. in particular, the forced basic typecheck on every option (even without calling `merge` or using the option) could potentially evaluate more than we'd like. this expression demonstrates what i mean by this: `builtins.typeOf ([] ++ (if builtins.trace "hello" true then ["example"] else []))`

consider this: it totally makes sense to insert a partial KDL document into a KDL document! that's quite useful, actually. in many parts of my niri settings rendering code, i render out a "partial document": e.g. the entire `input` section is just a node, but the `spawn-at-startup` "section" is actually arbitrarily many nodes; both of these are partial documents, and the latter *has* to be a list in and of itself. in my code, [i wrap the entire document in a "normalization" function that flattens the entire document](https://github.com/sodiboo/niri-flake/blob/1663b8f450988a64d9777f377efbd2fd8f3f243b/settings.nix#L3231) for precisely this purpose (since Nix has no "splat" syntax). [i even reimplement a `mkIf`-equivalent](https://github.com/sodiboo/niri-flake/blob/1663b8f450988a64d9777f377efbd2fd8f3f243b/settings.nix#L3240).

i *think* these kinds of operations will be quite common, so what if we just support them first-class? well, basically the changes needed to do so are quite minimal: the "KDL document" type is now more like `(types.listOf(...)).merge`; it will expand `lib.mkIf`, but because it knows *exactly* what its "element" type is, it can recursively descend down all "lists" (partial documents) until it gets a flattened list of nodes. This is a lot nicer to write complex rendering code for ("rendering" == "turning Nix options into a KDL document"). [Here's where i did that](https://github.com/sodiboo/nixpkgs/commit/f72a3572e07b2c8a9b2577093083746ea75bcbd8). this solves the strictness issues with making the toplevel type `mergeUniqueOption`.
